File: shared/src/main/java/com/runtracker/shared/data/repository/CyclingRepository.kt
package com.runtracker.shared.data.repository

import com.runtracker.shared.data.db.CyclingWorkoutDao
import com.runtracker.shared.data.db.CyclingTrainingPlanDao
import com.runtracker.shared.data.model.*
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import java.util.*
class CyclingRepository(
    private val workoutDao: CyclingWorkoutDao,
    private val planDao: CyclingTrainingPlanDao
) {
    // Workout operations
    fun getAllWorkouts(): Flow<List<CyclingWorkout>> = workoutDao.getAllWorkouts()
    
    suspend fun getWorkoutById(id: Long): CyclingWorkout? = workoutDao.getWorkoutById(id)
    
    fun getWorkoutsInRange(startTime: Long, endTime: Long): Flow<List<CyclingWorkout>> = 
        workoutDao.getWorkoutsInRange(startTime, endTime)
    
    fun getRecentCompletedWorkouts(limit: Int = 10): Flow<List<CyclingWorkout>> = 
        workoutDao.getRecentCompletedWorkouts(limit)
    
    suspend fun getActiveWorkout(): CyclingWorkout? = workoutDao.getActiveWorkout()
    
    suspend fun insertWorkout(workout: CyclingWorkout): Long = workoutDao.insert(workout)
    
    suspend fun updateWorkout(workout: CyclingWorkout) = workoutDao.update(workout)
    
    suspend fun deleteWorkout(workout: CyclingWorkout) = workoutDao.delete(workout)
    
    suspend fun getTotalDistanceSince(startTime: Long): Double = 
        workoutDao.getTotalDistanceSince(startTime) ?: 0.0
    
    suspend fun getWorkoutCountSince(startTime: Long): Int = 
        workoutDao.getWorkoutCountSince(startTime)
    
    suspend fun getAveragePowerSince(startTime: Long): Double = 
        workoutDao.getAveragePowerSince(startTime) ?: 0.0
    
    // Training plan operations
    fun getAllPlans(): Flow<List<CyclingTrainingPlan>> = planDao.getAllPlans()
    
    fun getActivePlan(): Flow<CyclingTrainingPlan?> = planDao.getActivePlan()
    
    suspend fun getPlanById(id: Long): CyclingTrainingPlan? = planDao.getPlanById(id)
    
    suspend fun insertPlan(plan: CyclingTrainingPlan): Long = planDao.insert(plan)
    
    suspend fun updatePlan(plan: CyclingTrainingPlan) = planDao.update(plan)
    
    suspend fun deletePlan(plan: CyclingTrainingPlan) = planDao.delete(plan)
    
    suspend fun activatePlan(id: Long) {
        planDao.deactivateAllPlans()
        planDao.activatePlan(id)
    }
    
    // Get next upcoming workout from active plan
    suspend fun getNextUpcomingWorkout(): Pair<ScheduledCyclingWorkout, Int>? {
        val plan = getActivePlan().first() ?: return null
        
        val calendar = Calendar.getInstance()
        val startDate = Calendar.getInstance().apply { timeInMillis = plan.startDate }
        
        // Find next workout from today onwards
        for (daysAhead in 0..60) {
            val checkDate = Calendar.getInstance().apply { add(Calendar.DAY_OF_YEAR, daysAhead) }
            val checkDayOfWeek = checkDate.get(Calendar.DAY_OF_WEEK)
            val checkWeek = ((checkDate.timeInMillis - startDate.timeInMillis) / (7 * 24 * 60 * 60 * 1000)).toInt() + 1
            
            val workout = plan.weeklySchedule.find { 
                it.weekNumber == checkWeek && 
                it.dayOfWeek == checkDayOfWeek && 
                !it.isCompleted &&
                it.workoutType != CyclingWorkoutType.REST_DAY
            }
            
            if (workout != null) {
                return Pair(workout, daysAhead)
            }
        }
        
        return null
    }
    
    // Mark workout as completed
    suspend fun markWorkoutCompleted(planId: Long, workoutId: String, completedWorkoutId: Long) {
        val plan = getPlanById(planId) ?: return
        val updatedSchedule = plan.weeklySchedule.map { workout ->
            if (workout.id == workoutId) {
                workout.copy(isCompleted = true, completedWorkoutId = completedWorkoutId)
            } else {
                workout
            }
        }
        updatePlan(plan.copy(weeklySchedule = updatedSchedule))
    }
    
    // Calculate weekly stats
    suspend fun getWeeklyStats(): CyclingWeeklyStats {
        val calendar = Calendar.getInstance()
        calendar.set(Calendar.DAY_OF_WEEK, calendar.firstDayOfWeek)
        calendar.set(Calendar.HOUR_OF_DAY, 0)
        calendar.set(Calendar.MINUTE, 0)
        calendar.set(Calendar.SECOND, 0)
        calendar.set(Calendar.MILLISECOND, 0)
        val weekStart = calendar.timeInMillis
        
        val totalDistance = getTotalDistanceSince(weekStart)
        val workoutCount = getWorkoutCountSince(weekStart)
        val avgPower = getAveragePowerSince(weekStart)
        
        return CyclingWeeklyStats(
            totalDistanceMeters = totalDistance,
            workoutCount = workoutCount,
            averagePowerWatts = avgPower,
            weekStartTimestamp = weekStart
        )
    }
}

data class CyclingWeeklyStats(
    val totalDistanceMeters: Double,
    val workoutCount: Int,
    val averagePowerWatts: Double,
    val weekStartTimestamp: Long
) {
    val totalDistanceKm: Double get() = totalDistanceMeters / 1000.0
}

File: iosApp/GoSteady/Sources/Repositories/CyclingRepository.swift
import Foundation
import SwiftData

final class CyclingRepository {
    private let context: ModelContext

    init(context: ModelContext) {
        self.context = context
    }

    // MARK: - Workouts

    func fetchAllWorkouts() -> [CyclingWorkout] {
        let descriptor = FetchDescriptor<CyclingWorkout>(sortBy: [SortDescriptor(\.startTime, order: .reverse)])
        return (try? context.fetch(descriptor)) ?? []
    }

    func fetchWorkoutById(_ id: UUID) -> CyclingWorkout? {
        let descriptor = FetchDescriptor<CyclingWorkout>(predicate: #Predicate { $0.id == id })
        return try? context.fetch(descriptor).first
    }

    func fetchWorkoutsInRange(start: Date, end: Date) -> [CyclingWorkout] {
        let descriptor = FetchDescriptor<CyclingWorkout>(
            predicate: #Predicate { $0.startTime >= start && $0.startTime <= end },
            sortBy: [SortDescriptor(\.startTime, order: .reverse)]
        )
        return (try? context.fetch(descriptor)) ?? []
    }

    func fetchRecentCompleted(limit: Int = 10) -> [CyclingWorkout] {
        var descriptor = FetchDescriptor<CyclingWorkout>(
            predicate: #Predicate { $0.isCompleted == true },
            sortBy: [SortDescriptor(\.startTime, order: .reverse)]
        )
        descriptor.fetchLimit = limit
        return (try? context.fetch(descriptor)) ?? []
    }

    func fetchActiveWorkout() -> CyclingWorkout? {
        let descriptor = FetchDescriptor<CyclingWorkout>(predicate: #Predicate { $0.isCompleted == false })
        return try? context.fetch(descriptor).first
    }

    func insert(_ workout: CyclingWorkout) {
        context.insert(workout)
        try? context.save()
    }

    func update(_ workout: CyclingWorkout) {
        try? context.save()
    }

    func delete(_ workout: CyclingWorkout) {
        context.delete(workout)
        try? context.save()
    }

    func totalDistanceSince(_ date: Date) -> Double {
        fetchWorkoutsInRange(start: date, end: Date())
            .filter(\.isCompleted)
            .reduce(0) { $0 + $1.distanceMeters }
    }

    func workoutCountSince(_ date: Date) -> Int {
        fetchWorkoutsInRange(start: date, end: Date())
            .filter(\.isCompleted)
            .count
    }

    func averagePowerSince(_ date: Date) -> Double {
        let workouts = fetchWorkoutsInRange(start: date, end: Date())
            .filter(\.isCompleted)
            .compactMap(\.avgPowerWatts)
        guard !workouts.isEmpty else { return 0 }
        return Double(workouts.reduce(0, +)) / Double(workouts.count)
    }

    // MARK: - Training Plans

    func fetchAllPlans() -> [CyclingTrainingPlan] {
        let descriptor = FetchDescriptor<CyclingTrainingPlan>(sortBy: [SortDescriptor(\.startDate, order: .reverse)])
        return (try? context.fetch(descriptor)) ?? []
    }

    func fetchActivePlan() -> CyclingTrainingPlan? {
        let descriptor = FetchDescriptor<CyclingTrainingPlan>(predicate: #Predicate { $0.isActive == true })
        return try? context.fetch(descriptor).first
    }

    func fetchPlanById(_ id: UUID) -> CyclingTrainingPlan? {
        let descriptor = FetchDescriptor<CyclingTrainingPlan>(predicate: #Predicate { $0.id == id })
        return try? context.fetch(descriptor).first
    }

    func insertPlan(_ plan: CyclingTrainingPlan) {
        context.insert(plan)
        try? context.save()
    }

    func updatePlan(_ plan: CyclingTrainingPlan) {
        try? context.save()
    }

    func deletePlan(_ plan: CyclingTrainingPlan) {
        context.delete(plan)
        try? context.save()
    }

    func activatePlan(_ plan: CyclingTrainingPlan) {
        let allPlans = fetchAllPlans().filter(\.isActive)
        for p in allPlans { p.isActive = false }
        plan.isActive = true
        try? context.save()
    }

    func markWorkoutCompleted(planId: UUID, workoutId: String, completedWorkoutId: UUID) {
        guard let plan = fetchPlanById(planId) else { return }
        if let index = plan.weeklySchedule.firstIndex(where: { $0.id == workoutId }) {
            plan.weeklySchedule[index].isCompleted = true
            plan.weeklySchedule[index].completedWorkoutId = completedWorkoutId
            try? context.save()
        }
    }

    // MARK: - Weekly Stats

    func weeklyStats() -> CyclingWeeklyStats {
        let weekStart = Date().startOfWeek
        let totalDistance = totalDistanceSince(weekStart)
        let count = workoutCountSince(weekStart)
        let avgPower = averagePowerSince(weekStart)
        return CyclingWeeklyStats(
            totalDistanceMeters: totalDistance,
            workoutCount: count,
            averagePowerWatts: avgPower,
            weekStartDate: weekStart
        )
    }
}

struct CyclingWeeklyStats {
    let totalDistanceMeters: Double
    let workoutCount: Int
    let averagePowerWatts: Double
    let weekStartDate: Date

    var totalDistanceKm: Double { totalDistanceMeters / 1000.0 }
}
